File: frontend/iosApp/UILogics/Sources/UILogics/Screens/SplashViewModel.swift
import Foundation
import Domains
import UseCases

@MainActor
public final class SplashViewModel: ObservableObject {
    @Published public var state: State = .initial

    private let splashUseCase: SplashUseCaseProtocol
    private let onSuccess: (User) -> Void

    public init(
        splashUseCase: SplashUseCaseProtocol,
        onSuccess: @escaping (User) -> Void
    ) {
        self.splashUseCase = splashUseCase
        self.onSuccess = onSuccess
    }

    public func launchApp() async {
        state = .loading(message: "Loading user data...")

        let result = await splashUseCase.launchApp()

        switch result {
        case .success(let user):
            onSuccess(user)
        case .failure(let error):
            handleError(error)
        }
    }

    private func handleError(_ error: SplashUseCaseModel.LaunchAppResult.Error) {
        switch error {
        case .sessionExpired:
            state = .error(.init(
                icon: "clock.badge.exclamationmark",
                iconColor: .red,
                message: "Session has expired",
                buttonTitle: "Go to Login",
                action: { [weak self] in
                    self?.splashUseCase.clearSession()
                }
            ))
        case .networkError:
            state = .error(.init(
                icon: "exclamationmark.triangle",
                iconColor: .orange,
                message: "Network error occurred",
                buttonTitle: "Retry",
                action: { [weak self] in
                    Task { await self?.launchApp() }
                }
            ))
        case .offlineNoCache:
            state = .error(.init(
                icon: "wifi.slash",
                iconColor: .orange,
                message: "You're offline with no cached data",
                buttonTitle: "Retry",
                action: { [weak self] in
                    Task { await self?.launchApp() }
                }
            ))
        case .serverError:
            state = .error(.init(
                icon: "exclamationmark.triangle",
                iconColor: .orange,
                message: "Server error occurred",
                buttonTitle: "Retry",
                action: { [weak self] in
                    Task { await self?.launchApp() }
                }
            ))
        case .unknown:
            state = .error(.init(
                icon: "exclamationmark.triangle",
                iconColor: .orange,
                message: "Unknown error occurred",
                buttonTitle: "Retry",
                action: { [weak self] in
                    Task { await self?.launchApp() }
                }
            ))
        }
    }
}

extension SplashViewModel {
    public enum State {
        case initial
        case loading(message: String)
        case error(ErrorItem)

        public struct ErrorItem {
            public let icon: String
            public let iconColor: IconColor
            public let message: String
            public let buttonTitle: String
            public let action: () -> Void

            public enum IconColor {
                case red
                case orange
            }
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/SplashViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.domain.User
import com.example.memoriesapp.usecase.LaunchAppError
import com.example.memoriesapp.usecase.LaunchAppResult
import com.example.memoriesapp.usecase.SplashUseCase
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for splash screen.
 */
class SplashViewModel(
    private val splashUseCase: SplashUseCase
) : ViewModel() {

    var state by mutableStateOf<State>(State.Initial)
        private set

    private val _navigationEvent = MutableSharedFlow<NavigationEvent>()
    val navigationEvent: SharedFlow<NavigationEvent> = _navigationEvent.asSharedFlow()

    fun launchApp() {
        viewModelScope.launch {
            state = State.Loading("Loading user data...")
            when (val result = splashUseCase.launchApp()) {
                is LaunchAppResult.Success -> {
                    _navigationEvent.emit(NavigationEvent.LaunchSuccess(result.user))
                }
                is LaunchAppResult.Failure -> {
                    handleError(result.error)
                }
            }
        }
    }

    private fun handleError(error: LaunchAppError) {
        state = when (error) {
            LaunchAppError.SESSION_EXPIRED -> State.Error(
                ErrorItem(
                    message = "Session has expired",
                    buttonTitle = "Go to Login",
                    action = {
                        splashUseCase.clearSession()
                        viewModelScope.launch {
                            _navigationEvent.emit(NavigationEvent.SessionExpired)
                        }
                    }
                )
            )
            LaunchAppError.NETWORK_ERROR -> State.Error(
                ErrorItem(
                    message = "Network error occurred",
                    buttonTitle = "Retry",
                    action = { launchApp() }
                )
            )
            LaunchAppError.OFFLINE_NO_CACHE -> State.Error(
                ErrorItem(
                    message = "You're offline with no cached data",
                    buttonTitle = "Retry",
                    action = { launchApp() }
                )
            )
            LaunchAppError.SERVER_ERROR -> State.Error(
                ErrorItem(
                    message = "Server error occurred",
                    buttonTitle = "Retry",
                    action = { launchApp() }
                )
            )
            LaunchAppError.UNKNOWN -> State.Error(
                ErrorItem(
                    message = "Unknown error occurred",
                    buttonTitle = "Retry",
                    action = { launchApp() }
                )
            )
        }
    }

    sealed class State {
        data object Initial : State()
        data class Loading(val message: String) : State()
        data class Error(val item: ErrorItem) : State()
    }

    data class ErrorItem(
        val message: String,
        val buttonTitle: String,
        val action: () -> Unit
    )

    sealed class NavigationEvent {
        data class LaunchSuccess(val user: User) : NavigationEvent()
        data object SessionExpired : NavigationEvent()
    }
}
